elasticsearch_dsl/response/__init__.py

```diff
@@ -15,61 +15,89 @@
 #  specific language governing permissions and limitations
 #  under the License.
 
-from typing import TYPE_CHECKING
-
-from ..utils import AttrDict, AttrList, _wrap
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    Generic,
+    Iterator,
+    List,
+    Optional,
+    Tuple,
+    Union,
+    cast,
+)
+
+from typing_extensions import TypeVar
+
+from ..utils import AttrDict, AttrList, JSONType, _wrap
 from .hit import Hit, HitMeta
 
 if TYPE_CHECKING:
-    from ..search_base import SearchBase
+    from ..aggs import Agg
+    from ..search_base import Request, SearchBase
 
 __all__ = ["Response", "AggResponse", "UpdateByQueryResponse", "Hit", "HitMeta"]
 
+_R = TypeVar("_R", default=Hit)
+
 
-class Response(AttrDict):
-    def __init__(self, search, response, doc_class=None):
+class Response(AttrDict[JSONType], Generic[_R]):
+    if TYPE_CHECKING:
+        _search: "SearchBase[_R]"
+        _doc_class: Optional[_R]
+        _hits: List[_R]
+
+    def __init__(
+        self,
+        search: "Request[_R]",
+        response: Dict[str, Any],
+        doc_class: Optional[_R] = None,
+    ):
         super(AttrDict, self).__setattr__("_search", search)
         super(AttrDict, self).__setattr__("_doc_class", doc_class)
         super().__init__(response)
 
-    def __iter__(self):
+    def __iter__(self) -> Iterator[_R]:  # type: ignore[override]
         return iter(self.hits)
 
-    def __getitem__(self, key):
+    def __getitem__(self, key: Union[slice, int, str]) -> Any:
         if isinstance(key, (slice, int)):
             # for slicing etc
             return self.hits[key]
         return super().__getitem__(key)
 
-    def __nonzero__(self):
+    def __nonzero__(self) -> bool:
         return bool(self.hits)
 
     __bool__ = __nonzero__
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return "<Response: %r>" % (self.hits or self.aggregations)
 
-    def __len__(self):
+    def __len__(self) -> int:
         return len(self.hits)
 
-    def __getstate__(self):
+    def __getstate__(self) -> Tuple[Dict[str, Any], "SearchBase[_R]", Optional[_R]]:  # type: ignore[override]
         return self._d_, self._search, self._doc_class
 
-    def __setstate__(self, state):
+    def __setstate__(
+        self, state: Tuple[Dict[str, Any], "SearchBase[_R]", Optional[_R]]  # type: ignore[override]
+    ) -> None:
         super(AttrDict, self).__setattr__("_d_", state[0])
         super(AttrDict, self).__setattr__("_search", state[1])
         super(AttrDict, self).__setattr__("_doc_class", state[2])
 
-    def success(self):
+    def success(self) -> bool:
         return self._shards.total == self._shards.successful and not self.timed_out
 
     @property
-    def hits(self):
+    def hits(self) -> List[_R]:
         if not hasattr(self, "_hits"):
-            h = self._d_["hits"]
+            h = cast(AttrDict[JSONType], self._d_["hits"])
```

Comment:
This is an common issue in many places that I think comes from a not great `Dict[str, JSONType]` typing that we use for data coming from the API. I'm not sure if there is a better way to type this so that it does not require casting each time an item is extracted, at some point I will think about that.